### PR TITLE
6425: Fail to read a data item after data source compaction was interrupted

### PR DESCRIPTION
**Pull request by artemananiev**
Fix summary:

* A new flag is introduced to `DataFileReader` to indicate whether it's open or not
* The flag is used in `isOpen()` instead of checking a (random) file channel, which may happen to be closed by interrupt
* If a reader is not closed, the channel will be re-created once any thread tries to read using it, this code isn't touched in this fix

Fixes: https://github.com/hashgraph/hedera-services/issues/6425
Signed-off-by: Artem Ananev <artem.ananev@swirldslabs.com>

### CODE REVIEW

File: platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/ImmutableIndexedObjectListUsingArray.java [hunk 1]
```diff
@@ -107,9 +108,10 @@ public ImmutableIndexedObjectListUsingArray<T> withAddedObject(final T newObject
     }
 
     /** {@inheritDoc} */
-    public ImmutableIndexedObjectListUsingArray<T> withDeletedObjects(final Set<T> objectsToDelete) {
+    @Override
+    public ImmutableIndexedObjectListUsingArray<T> withDeletedObjects(@NonNull final Collection<T> objectsToDelete) {
         // Ignore null objects, share an immutable empty list
```
Comment (OlegMazurov) [inline platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/ImmutableIndexedObjectListUsingArray.java:113]: Fix this comment?

**Review by OlegMazurov — DISMISSED** [review]

File: platform-sdk/swirlds-jasperdb/src/test/java/com/swirlds/merkledb/files/DataFileReaderCloseTest.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                    Assertions.assertEquals(i, item[0]);
+                    Assertions.assertEquals(i + 1, item[1]);
+                    readingThreadStarted.set(true);
+                }
+            } catch (final ClosedByInterruptException e) {
+                // This is expected
+            } catch (final IOException e) {
+                exceptionOccurred.set(e);
+            }
+            readingThreadInterrupted.set(true);
```
Comment (jsync-swirlds) [inline platform-sdk/swirlds-jasperdb/src/test/java/com/swirlds/merkledb/files/DataFileReaderCloseTest.java:82]: should this line be in the catch for closed by interrupted exception?

File: platform-sdk/swirlds-jasperdb/src/test/java/com/swirlds/merkledb/collections/ImmutableIndexedObjectListUsingArrayTest.java [hunk 3]
```diff
@@ -111,7 +111,6 @@ void emptyInputYieldsEmptyListWithNoopDeletion() {
                 factoryForReaderToTest(new TestIndexedObject[0]);
 
         assertNull(emptyList.getLast(), "Last element of empty list should be null");
-        assertSame(emptyList, emptyList.withDeletedObjects(null), "Lists should reuse themselves given null deletions");
```
[the remark this answers is not in the document]
Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/test/java/com/swirlds/merkledb/collections/ImmutableIndexedObjectListUsingArrayTest.java:114]: This is already done in the test, right below.

File: platform-sdk/swirlds-jasperdb/src/test/java/com/swirlds/merkledb/files/DataFileReaderCloseTest.java (hunk 2 above)
Comment (artemananiev) [reply platform-sdk/swirlds-jasperdb/src/test/java/com/swirlds/merkledb/files/DataFileReaderCloseTest.java:82]: I think the var name is confusing, and the var seems redundant at all. Let me re-implement it using `Thread.join()` instead.

**Review by OlegMazurov — APPROVED** [review]

**Review by jsync-swirlds — APPROVED** [review]

**Pull request merged**